honeybadgermpc/protocols/commoncoin.py

```diff
@@ -92,7 +93,85 @@ def hash(x):
         logging.debug(
                       f"broadcast {('COIN', round, sk.sign(h))}",
                       extra={'nodeid': pid, 'epoch': round})
-        broadcast(('COIN', round, sk.sign(h)))
+        sign_str = codecs.encode(serialize(sk.sign(h)), "base64").decode()
+        broadcast(('COIN', round, sign_str))
         return await output_queue[round].get()
 
     return get_coin, recv_task
+
+
+async def run_common_coin(config, pbk, pvk, n, f, nodeid):
+    program_runner = ProcessProgramRunner(config, n, t, nodeid)
+    sender, listener = program_runner.senders, program_runner.listener
+
+    await sender.connect()
+
+    _, recv, broadcast = program_runner.get_send_and_recv("coin")
+    coin, crecv_task = await shared_coin('sidA', nodeid, n, f, pbk, pvk, broadcast, recv)
+    for i in range(10):
+        logging.info("%d COIN VALUE: %s", i, await coin(i))
+    crecv_task.cancel()
+
+    await sender.close()
+    await listener.close()
+
+
+if __name__ == "__main__":
+    import os
+    import sys
+    import pickle
+    from honeybadgermpc.exceptions import ConfigurationError
+    from honeybadgermpc.config import load_config
+    from honeybadgermpc.ipc import NodeDetails, ProcessProgramRunner
+    from honeybadgermpc.protocols.crypto.boldyreva import TBLSPublicKey  # noqa:F401
+    from honeybadgermpc.protocols.crypto.boldyreva import TBLSPrivateKey  # noqa:F401
+
+    configfile = os.environ.get('HBMPC_CONFIG')
+    nodeid = os.environ.get('HBMPC_NODE_ID')
+    pvk_string = os.environ.get('HBMPC_PV_KEY')
+    pbk_string = os.environ.get('HBMPC_PB_KEY')
+
+    # override configfile if passed to command
+    try:
+        nodeid = sys.argv[1]
+        configfile = sys.argv[2]
+        pbk_string = sys.argv[3]
+        pvk_string = sys.argv[4]
+    except IndexError:
+        pass
+
+    if not nodeid:
+        raise ConfigurationError('Environment variable `HBMPC_NODE_ID` must be set'
+                                 ' or a node id must be given as first argument.')
+
+    if not configfile:
+        raise ConfigurationError('Environment variable `HBMPC_CONFIG` must be set'
+                                 ' or a config file must be given as first argument.')
+
+    if not pvk_string:
+        raise ConfigurationError('Environment variable `HBMPC_PV_KEY` must be set'
+                                 ' or a config file must be given as first argument.')
+
+    if not pbk_string:
+        raise ConfigurationError('Environment variable `HBMPC_PB_KEY` must be set'
+                                 ' or a config file must be given as first argument.')
+
+    config_dict = load_config(configfile)
+    n = config_dict['N']
+    t = config_dict['t']
+    k = config_dict['k']
+    pbk = pickle.loads(codecs.decode(pbk_string.encode(), "base64"))
```

Comment:
this looks weird, we seem to be encoding and decoding in the same line, base64 and pickle are involved, can this be simplified?

Reply:
I will change this to `pickle.loads(base64.b64decode(pbk_string))`. The base64 encoding makes it convenient to store the pickled string in the shell script to run it. Otherwise, it is a pain to escape the characters within the pickled string to store them in the shell script.

Reply:
This has also been updated.